tests/core/test_variable.py

```diff
@@ -0,0 +1,771 @@
+    return (function.zeros() + 1) * number
+
+
+def has_dated_attribute(variable):
+    return variable.end is not None
+
+
+def add_variable_catch_assertion(tax_benefit_system, variable, assertion_message_prefix):
+    try:
+        tax_benefit_system.add_variable(variable)
+    except AssertionError, e:
+        if hasattr(e, 'message'):
+            message = getattr(e, 'message')
+            assert message.startswith(assertion_message_prefix), message
+        else:
+            raise
+    except:
+        raise
+
+
+# TESTS
+
+
+# NO DATED ATTRIBUTE(S) - NO DATED FORMULA
+
+
+class variable__no_dated_attributes(Variable):
+    column = IntCol
+    entity = Individu
+    definition_period = MONTH
+    label = u"Variable without date."
+
+
+tax_benefit_system.add_variable(variable__no_dated_attributes)
+
+
+def test_variable__no_dated_attributes():
+    variable = tax_benefit_system.column_by_name['variable__no_dated_attributes']
+    assert variable is not None
+    assert variable.end is None
+    assert variable.formula_class.dated_formulas_class.__len__() == 0
+
+
+# DATED ATTRIBUTE(S) - NO DATED FORMULA
+
+
+class variable__deprecated_start_date(Variable):
+    column = IntCol
+    entity = Individu
+    definition_period = MONTH
+    label = u"Variable with dated attributes (including deprecated start_date), no function."
+    start_date = datetime.date(1980, 1, 1)  # Deprecated
+    end = '1989-12-31'
+
+
+add_variable_catch_assertion(tax_benefit_system, variable__deprecated_start_date, 'Deprecated "start_date" attribute in definition of variable')
+
+
+# 371 - end, no function
+
+class variable__strange_end_attribute(Variable):
+    column = IntCol
+    entity = Individu
+    definition_period = MONTH
+    label = u"Variable with dubious end attribute, no function."
+    end = '1989-00-00'
+
+
+def test_variable__strange_end_attribute():
+    try:
+        tax_benefit_system.add_variable(variable__strange_end_attribute)
+
+    except ValueError, e:
+        if hasattr(e, 'message'):
+            message = getattr(e, 'message')
+            assert message.startswith("Incorrect 'end' attribute format"), message
+    except:
+        raise
+
+    # Check that Error at variable adding prevents it from registration in the taxbenefitsystem.
+    assert not hasattr(tax_benefit_system.column_by_name, "variable__strange_end_attribute")
+
+
+# 371 - end, no function
+
+class variable__dated_attribute(Variable):
+    column = IntCol
+    entity = Individu
+    definition_period = MONTH
+    label = u"Variable with end attribute, no function."
+    # start_date = datetime.date(1980, 1, 1)
+    end = '1989-12-31'
+
+
+tax_benefit_system.add_variable(variable__dated_attribute)
+
+
+def test_variable__dated_attribute():
+    variable = tax_benefit_system.column_by_name['variable__dated_attribute']
+    assert variable is not None
+    assert variable.end == '1989-12-31'
+    assert variable.formula_class.dated_formulas_class.__len__() == 0
```

Comment:
It's not about testing all combinations : an end date changes how the validity of a function is estimated. But, it's tested elsewhere so ok for removing it.